lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+
+/* Internal variable used to generate message IDs. */
+static uint16_t message_id_next = QOS_MESSAGE_ID_BASE;
+
+/* Internal API functions */
+
+static void notify_event(const struct qos_evt *evt)
+{
+	__ASSERT(evt != NULL, "Library event not found");
+
+	if (ctx.app_evt_handler != NULL) {
+		ctx.app_evt_handler(evt);
+	}
+}
+
+static int timeout_get(void)
+{
+	if (ctx.retry_count > ARRAY_SIZE(backoff_delay_lookup_sec) - 1) {
+		struct qos_evt evt = {
+			.type = QOS_EVT_RETRY_COUNT_EXPIRED
+		};
+		notify_event(&evt);
+
+		/* Reset count after notification. */
+		ctx.retry_count = 0;
+		return -ERANGE;
+	}
+
+	return backoff_delay_lookup_sec[ctx.retry_count];
+}
+
+static int timer_start(void)
+{
+	if (!k_work_delayable_is_pending(&ctx.timeout_handler_work)) {
+		int timeout = timeout_get();
+
+		LOG_DBG("Timeout until next notification: %d", timeout);
+
+		if (timeout < 0) {
+			LOG_WRN("Maximum retries reached, abort message notification");
+			return -ERANGE;
+		}
+
+		k_work_reschedule(&ctx.timeout_handler_work, K_SECONDS(timeout));
+	}
+
+	return 0;
+}
+
+static void timeout_handler_work_fn(struct k_work *work)
+{
+	int timeout;
+
+	/* Notify all messages that are present in internal pending list. */
+	qos_message_notify_all();
+
+	timeout = timeout_get();
+	if (timeout < 0) {
+		LOG_WRN("Maximum retries reached, abort message notification");
+		return;
+	}
+
+	k_work_reschedule(&ctx.timeout_handler_work, K_SECONDS(timeout));
+}
+
+/* @brief Function that appends a message to the internal list of pending messages.
+ *
+ * @returns A positive value indicating the index of the internal list that the message was
+ *	    added to. Otherwise a negative value indicating the reason of failure.
+ *
+ * @retval -ENOMEM if the internal list is full.
+ */
+static int list_append(struct qos_data *message)
+{
+	int index_next = INVALID;
+
+	/* Find the next available list index. */
+	for (int i = 0; i < ARRAY_SIZE(ctx.list_internal) - 1; i++) {
+		if (ctx.list_internal[i].message.notified_count == 0) {
+			index_next = i;
+			break;
+		}
+	}
+
+	if (index_next == INVALID) {
+		LOG_ERR("No available entries in pending message list");
+		return -ENOMEM;
+	}
+
+	ctx.list_internal[index_next].message = *message;
+
+	sys_slist_append(&ctx.pending_list, &ctx.list_internal[index_next].header);
+	return index_next;
+}
+
+static int list_remove(uint32_t id)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL, *node_prev = NULL;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST,
+	};
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		if (node->message.id == id) {
+			evt.message = node->message;
+			notify_event(&evt);
+			memset(&node->message, 0, sizeof(struct qos_data));
+			sys_slist_remove(&ctx.pending_list, &node_prev->header, &node->header);
+			return 0;
+		}
+
+		node_prev = node;
+	};
+
+	return -ENODATA;
+}
+
+/* Public API functions */
+
+int qos_init(qos_evt_handler_t evt_handler)
```

Comment:
A single event handler function only allows for a single user of the library. Up to you really, but wouldn't it be better to  keep cb function inside `struct qos_data`?

Reply:
I think I'll leave it as is for now. We can always add this later on if a need presents it self.

Reply:
Sounds like a good future improvement to allow for more than one user